### ISSUE TITLE
Adding the visuallyhidden class to a table caption does not completely remove it from layout

### ISSUE THREAD

**Issue by mattbrundage** [h5bp/html5-boilerplate#1932]
Adding the `visuallyhidden` class to a table caption does not completely remove it from layout.

Reduced test case: https://codepen.io/mattbrundage/pen/evPZNe
Demonstrated in Firefox, Chrome, IE 11, and Edge.

@ffoodd 
@roblarsen

**Comment by ffoodd** [comment]
Thanks for the reduced test case!

First of all, I think the explanation is the table layout model: it creates spacings between borders… Just try a background on cells to show those spacings. 

These spacings exists even if you remove every borders. That's where `.visuallyhidden` on `<caption>` won't remove it from layout: despite the `border: 0;` thing, the able layout still creates space around it.

Let me suggest two ways to get around this:

1. if possible, use `border-collapse: collapse;` or `border-spacing: 0;` on `table` element: this might be troubling your layout so I think it's not the best way;
2. use a `<span>` inside your `<caption>` to contain your text, and hide the span. I think this is the best way.

Here's a pen forked from yours to show this: http://codepen.io/ffoodd/pen/evPjXY
I tested this on Firefox and Chrome (currently don't have any other browser to test) but that does the trick.

To sum up, there's nothing we can do in `.visuallyhidden` to improve this, since what's causing issue is the parent layout table…

Let me know if that solves your issue :)

**Comment by mattbrundage** [comment]
@ffoodd, excellent points; I agree that there is nothing we can do with `.visuallyhidden`.

**Your first suggestion** works only if there are no borders on `<th>` and `<td>` elements. When a table's `<caption>` is absolutely positioned (as with `.visuallyhidden`), it resides directly _below_ the `<thead>`, rather than above it. This prevents the borders between `<thead>` and `<tbody>` from collapsing into each other. Example: https://codepen.io/mattbrundage/pen/wJQagK

**Your second suggestion** appears to work, but I noticed that the `<caption>` ends up with a height of 0, which may cause screen readers to skip over it.

**Comment by ffoodd** [comment]
Good question. I'll try to gather some data on a zero height for a `<caption>`…

**Comment by ffoodd** [comment]
First tries on VoiceOVer iOS and NVDA + Firefox, `<caption>`'s content is read — which is good news :)

**Comment by mattbrundage** [comment]
That's promising. I wonder if VoiceOver's behavior has changed in this regard. Snook's research from 2011 revealed that "[Apple's Voice Over will not read content within an element that has zero height.](https://snook.ca/archives/html_and_css/hiding-content-for-accessibility)"

**Comment by ffoodd** [comment]
Indeed, this is the same technique we only improved, here.

But looking at our markup, the `<span>` in the `<caption>` is 1px × 1px.  But since it is out of the content flow (due to absolute positioning) it does not affect its parent's dimensions. For now it seems that it doesn't affect ATs, but needs a few more tests :)

**Comment by roblarsen** [comment]
Hi folks, is there anything we want to document about this issue? Otherwise, I'm closing.

**Comment by mattbrundage** [comment]
@roblarsen, you can close.
